# Laboratorio de regresión logística

|                |   |
:----------------|---|
| **Nombre**     |  Nicolás Martínez Gutiérrez |
| **Fecha**      |  13/10/2025 |
| **Expediente** |  751746 | 

In machine learning, Support Vector Machines (SVM) are supervised learning models with associated learning algorithms that analyze data used for classification and regression analysis. It is mostly used in classification problems. In this algorithm, each data item is plotted as a point in p-dimensional space (where p is the number of features), with the value of each feature being the value of a particular coordinate. Then, classification is performed by finding the hyper-plane that best differentiates the two classes (or more if we have a multi class problem):

$$ f(x) = w^T \varphi(x) + b $$

where $\varphi: X \rightarrow F $ is a function that makes each input point $x$ correspond to a point in F, where F is a Hilbert space.

In addition to performing linear classification, SVMs can efficiently perform a non-linear classification, implicitly mapping their inputs into high-dimensional feature spaces (more specifically using the kernel trick, like the RBF funcion). 

[1]

OLS utilizes the squared residuals to fit the parameters. Large residuals caused by outliers may worsen the accuracy significantly.

Support Vectors use piecewise linear functions to counter this, in which a hyperparameter  $\epsilon$ called the margin lets errors that are less or equal to it be 0, and error larger than it be $e - \epsilon$. 

The problem to solve is:

\begin{split}
        \min_{w, b, \xi, \xi^*} \mathcal{P}_\epsilon(w, b, \xi) &= \frac{1}{2} w^T w + c \sum_{k=1}^{N} \xi_k \\
        \text{s.t. } & y_k [w^T \varphi(x_k) - b] \geq 1- \xi_k,\ \ k = 1, ..., N \\
        & \xi_k \geq 0,\ \ k = 1, ..., N
\end{split}


The most important question that arises when using a SVM is how to choose the correct hyperplane. Consider the following scenarios:

### Scenario 1

In this scenario there are three hyperplanes called A, B, and C. Now, the problem is to identify the hyperplane which best differentiates the stars and the circles.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_21-2.png" alt="what image shows"></center>

In this case, hyperplane B separates the stars and the circle betters, hence it is the correct hyperplane.


### Scenario 2

Now take another scenario where all three hyperplanes are segregating classes well. The question that arises is how to choose the best hyperplane in this situation.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_4-2.png" alt="what image shows"></center>

In such scenarios, we calculate the margin (which is the distance between nearest data point and the hyperplane). The hyperplane with the largest margin will be considered as the correct hyperplane to classify the dataset.

Here C has the largest margin. Hence, it is considered as the best hyperplane.


### Kernels
Knowing 
$$ w = \sum_{k=1}^{N} \alpha_k y_k \varphi(x_k) $$

And
$$ y_{pred} = w^T \varphi(x) + b $$

Then 
$$ y_{pred} = (\sum_{k=1}^{N} \alpha_k y_k \varphi(x_k))^T \varphi(x) + b $$

Where $\varphi$ is a function that makes each input in $x$ correspond to a point in $F$ (a Hilbert space). This can be seen as processing and transforming the input featuers to keep the model's convexity. [2]

This also allows us to transform the inputs into another space where they might be more easily classified.

<center><img src=https://miro.medium.com/max/838/1*gXvhD4IomaC9Jb37tzDUVg.png alt="what image shows"></center>

## ROC and AUC

A ROC (Receiver Operating Characteristic) is a graph that shows how the classification model performs at the classification thresholds. 

ROC curves typically feature true positive rate on the Y axis, and false positive rate on the X axis. This means that the top left corner of the plot is the “ideal” point - a false positive rate of zero, and a true positive rate of one. This is not very realistic, but it does mean that a larger area under the curve (AUC) is usually better. [3]

True Positive Rate is a synonym for Recall and defined as:
$$ TPR = \frac{TP}{TP + FN} $$

False Positive Rate is a synonym for Specificity and defined as:

$$ FPR = \frac{FP}{FP + TN} $$

ROC curves are typically used in binary classification to study the output of a classifier. In order to extend ROC curve and ROC area to multi-label classification, it is necessary to binarize the output. One ROC curve can be drawn per label, but one can also draw a ROC curve by considering each element of the label indicator matrix as a binary prediction (micro-averaging).

E.g. If you lower a classification threshold, more items would be classified as positive, increasing False Positives and True Positives.

AUC stands for Area under the ROC.

## Ejercicio 1

- Utiliza el dataset `Iris`, modela con SVC y haz Cross-Validation de diferentes kernels ('linear', 'poly', 'rbf', 'sigmoid').
- Modela con LogisticRegression.
- El método de Cross-Validation es K-Folds con $k=10$.
- Utiliza el AUC como métrico de Cross-Validation.
- Compara resultados.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_iris
from sklearn.model_selection import StratifiedKFold, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression

In [2]:
X, y = load_iris(return_X_y=True)

In [3]:
cv = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)
m1, m2, m3 = "accuracy", "f1_macro", "roc_auc_ovr"

In [4]:
svc_linear = SVC(kernel="linear", probability=True, random_state=42)
svc_poly = SVC(kernel="poly", probability=True, random_state=42)
svc_rbf = SVC(kernel="rbf", probability=True, random_state=42)
svc_sigmoid = SVC(kernel="sigmoid", probability=True, random_state=42)
logreg = LogisticRegression(max_iter=2000, solver="lbfgs", multi_class="ovr", random_state=42)

In [5]:
acc_linear = cross_val_score(svc_linear, X, y, cv=cv, scoring=m1).mean()
acc_poly = cross_val_score(svc_poly, X, y, cv=cv, scoring=m1).mean()
acc_rbf = cross_val_score(svc_rbf, X, y, cv=cv, scoring=m1).mean()
acc_sigmoid = cross_val_score(svc_sigmoid, X, y, cv=cv, scoring=m1).mean()
acc_logreg = cross_val_score(logreg, X, y, cv=cv, scoring=m1).mean()

C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'mu

In [6]:
f1_linear = cross_val_score(svc_linear, X, y, cv=cv, scoring=m2).mean()
f1_poly = cross_val_score(svc_poly, X, y, cv=cv, scoring=m2).mean()
f1_rbf = cross_val_score(svc_rbf, X, y, cv=cv, scoring=m2).mean()
f1_sigmoid = cross_val_score(svc_sigmoid, X, y, cv=cv, scoring=m2).mean()
f1_logreg = cross_val_score(logreg, X, y, cv=cv, scoring=m2).mean()

C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'mu

In [7]:
auc_linear = cross_val_score(svc_linear, X, y, cv=cv, scoring=m3).mean()
auc_poly = cross_val_score(svc_poly, X, y, cv=cv, scoring=m3).mean()
auc_rbf = cross_val_score(svc_rbf, X, y, cv=cv, scoring=m3).mean()
auc_sigmoid = cross_val_score(svc_sigmoid, X, y, cv=cv, scoring=m3).mean()
auc_logreg = cross_val_score(logreg, X, y, cv=cv, scoring=m3).mean()

C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
C:\Users\Nicolas\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'mu

In [8]:
data = pd.DataFrame({
    "Model": ["SVC (linear)", "SVC (poly)", "SVC (rbf)", "SVC (sigmoid)", "Logistic Regression"],
    "Accuracy": [acc_linear, acc_poly, acc_rbf, acc_sigmoid, acc_logreg],
    "F1-macro": [f1_linear, f1_poly, f1_rbf, f1_sigmoid, f1_logreg],
    "ROC AUC (OvR)": [auc_linear, auc_poly, auc_rbf, auc_sigmoid, auc_logreg]
}).sort_values("ROC AUC (OvR)", ascending=False).reset_index(drop=True)

In [9]:
data

Model  Accuracy  F1-macro  ROC AUC (OvR)
0            SVC (rbf)  0.960000  0.959731       0.997333
1         SVC (linear)  0.980000  0.979377       0.996000
2           SVC (poly)  0.960000  0.959175       0.996000
3  Logistic Regression  0.940000  0.938552       0.986667
4        SVC (sigmoid)  0.073333  0.047778       0.746667

Los modelos SVC con kernel lineal y RBF fueron los que obtuvieron el mejor desempeño al clasificar el dataset Iris, con valores de accuracy de 0.98 y 0.96, F1-macro de 0.9794 y 0.9597, y ROC AUC (OvR) de 0.9960 y 0.9973 respectivamente. El SVC con kernel polinomial también tuvo un rendimiento muy alto, con un AUC de 0.9960, mientras que la regresión logística alcanzó resultados ligeramente menores, con un accuracy de 0.94, un F1-macro de 0.9386 y un AUC de 0.9867. En contraste, el SVC con kernel sigmoide tuvo un desempeño deficiente, con un accuracy de apenas 0.0733 y un AUC de 0.7467, mostrando que no logra separar bien las clases. En resumen, los kernels lineal, RBF y polinomial son los más efectivos para este dataset, mientras que el sigmoide no resulta adecuado.

## Ejercicio 2
- Repite el ejercicio 1 con el dataset `Default`. Utiliza `default` como target.

In [10]:
data = pd.read_csv("Default.csv")

In [11]:
y = data["default"].replace({"Yes": 1, "No": 0}).to_numpy()
X_num = (data[["balance", "income"]] - data[["balance", "income"]].mean()) / data[["balance", "income"]].std()
X_cat = pd.get_dummies(data[["student"]], drop_first=True)
X = pd.concat([X_num, X_cat], axis=1).to_numpy()

C:\Users\Nicolas\AppData\Local\Temp\ipykernel_12184\3771904528.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  y = data["default"].replace({"Yes": 1, "No": 0}).to_numpy()


In [12]:
cv = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)
m1, m2, m3 = "accuracy", "f1_macro", "roc_auc" 

In [13]:
svc_linear = SVC(kernel="linear", probability=True, class_weight="balanced", random_state=42)
svc_poly = SVC(kernel="poly", probability=True, class_weight="balanced", random_state=42)
svc_rbf = SVC(kernel="rbf", probability=True, class_weight="balanced", random_state=42)
svc_sigmoid = SVC(kernel="sigmoid", probability=True, class_weight="balanced", random_state=42)
logreg = LogisticRegression(max_iter=2000, solver="lbfgs", class_weight="balanced", random_state=42)

In [14]:
acc_linear = cross_val_score(svc_linear, X, y, cv=cv, scoring=m1).mean()
acc_poly = cross_val_score(svc_poly, X, y, cv=cv, scoring=m1).mean()
acc_rbf = cross_val_score(svc_rbf, X, y, cv=cv, scoring=m1).mean()
acc_sigmoid = cross_val_score(svc_sigmoid, X, y, cv=cv, scoring=m1).mean()
acc_logreg = cross_val_score(logreg, X, y, cv=cv, scoring=m1).mean()

In [15]:
f1_linear = cross_val_score(svc_linear, X, y, cv=cv, scoring=m2).mean()
f1_poly = cross_val_score(svc_poly, X, y, cv=cv, scoring=m2).mean()
f1_rbf = cross_val_score(svc_rbf, X, y, cv=cv, scoring=m2).mean()
f1_sigmoid = cross_val_score(svc_sigmoid, X, y, cv=cv, scoring=m2).mean()
f1_logreg = cross_val_score(logreg, X, y, cv=cv, scoring=m2).mean()

In [16]:
auc_linear = cross_val_score(svc_linear, X, y, cv=cv, scoring=m3).mean()
auc_poly = cross_val_score(svc_poly, X, y, cv=cv, scoring=m3).mean()
auc_rbf = cross_val_score(svc_rbf, X, y, cv=cv, scoring=m3).mean()
auc_sigmoid = cross_val_score(svc_sigmoid, X, y, cv=cv, scoring=m3).mean()
auc_logreg = cross_val_score(logreg, X, y, cv=cv, scoring=m3).mean()

In [19]:
results = pd.DataFrame({
    "Model": ["SVC (linear)", "SVC (poly)", "SVC (rbf)", "SVC (sigmoid)", "Logistic Regression"],
    "Accuracy": [acc_linear, acc_poly, acc_rbf, acc_sigmoid, acc_logreg],
    "F1-macro": [f1_linear, f1_poly, f1_rbf, f1_sigmoid, f1_logreg],
    "ROC AUC":  [auc_linear, auc_poly, auc_rbf, auc_sigmoid, auc_logreg]
}).sort_values("ROC AUC", ascending=False).reset_index(drop=True)

In [20]:
results

Model  Accuracy  F1-macro   ROC AUC
0  Logistic Regression    0.8655  0.616271  0.948297
1         SVC (linear)    0.8646  0.615288  0.948220
2           SVC (poly)    0.8822  0.629055  0.940819
3            SVC (rbf)    0.8540  0.605436  0.922427
4        SVC (sigmoid)    0.7061  0.478599  0.808810

Los resultados don el dataset Default muestran que la Regresión Logística fue el modelo con mejor desempeño general, con un accuracy de 0.8655, F1-macro de 0.6163 y AUC de 0.9483, lo que indica que logra clasificar correctamente la mayoría de los casos y mantiene un buen equilibrio entre precisión y recall. El SVC (lineal) obtuvo resultados muy similares (accuracy 0.8646, AUC 0.9482), lo que lo convierte en una alternativa casi igual de efectiva. El SVC (polinomial) también tuvo un rendimiento competitivo con un AUC de 0.9408, mientras que el SVC (RBF) fue un poco menos preciso (accuracy 0.854, AUC 0.9224). Por otro lado, el SVC (sigmoide) fue claramente el modelo menos eficiente, con un accuracy de 0.706 y un AUC de 0.8088. En resumen, los mejores modelos para este conjunto de datos fueron la Regresión Logística y el SVC lineal, que ofrecen la mejor combinación entre precisión y capacidad para distinguir correctamente los casos de incumplimiento.

# Addendum

Métricos disponibles para clasificación:
- ‘accuracy’
- ‘balanced_accuracy’
- ‘top_k_accuracy’
- ‘average_precision’
- ‘neg_brier_score’
- ‘f1’
- ‘f1_micro’
- ‘f1_macro’
- ‘f1_weighted’
- ‘f1_samples’
- ‘neg_log_loss’
- ‘precision’ etc.
- ‘recall’ etc.
- ‘jaccard’ etc.
- ‘roc_auc’
- ‘roc_auc_ovr’
- ‘roc_auc_ovo’
- ‘roc_auc_ovr_weighted’
- ‘roc_auc_ovo_weighted’
- ‘d2_log_loss_score’

# References

[1] Shigeo Abe.Support Vector Machines for Pattern Classification,2Ed.Springer-Verlag London,2010. ISBN978-1-84996-097-7. URLhttps://www.springer.com/gp/book/9781849960977.

[2] Johan A K Suykens, Tony Van Gestel, Jos De Brabanter, BartDe Moor, and Joos Vandewalle.Least Squares Support VectorMachines. World Scientific,2002. ISBN9789812381514. URLhttps://www.worldscientific.com/worldscibooks/10.1142/5089.

[3] Bradley, A. P. (1997). The use of the area under the ROC curve in the evaluation of machine learning algorithms. Pattern recognition, 30(7), 1145-1159. URL https://www.researchgate.net/post/how_can_I_interpret_the_ROC_curve_result